stan/math/rev/functor/dae.hpp

```diff
@@ -0,0 +1,175 @@
+#ifndef STAN_MATH_REV_FUNCTOR_DAE_HPP
+#define STAN_MATH_REV_FUNCTOR_DAE_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/functor/idas_integrator.hpp>
+#include <stan/math/rev/functor/dae_system.hpp>
+#include <ostream>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Solve the DAE initial value problem f(t, y, y')=0, y(t0) = yy0, y'(t0)=yp0 at
+ * a set of times, { t1, t2, t3, ... } using IDAS.
+ *
+ * \p f must define an operator() with the signature as:
+ *   template<typename T_yy, typename T_yp, typename... T_Args>
+ *   Eigen::Matrix<stan::return_type_t<T_yy, T_yp, T_Args...>, Eigen::Dynamic,
+ * 1> operator()(double t, const Eigen::Matrix<T_yy, Eigen::Dynamic, 1>& yy,
+ *     const Eigen::Matrix<T_yp, Eigen::Dynamic, 1>& yp,
+ *     std::ostream* msgs, const T_Args&... args);
+ *
+ * t is the time, yy the vector-valued state, yp the vector-valued
+ * state derivative, msgs a stream for error
+ * messages, and args the optional arguments passed to the DAE solve function
+ * (which are passed through to \p f without modification).
+ *
+ * @tparam F Type of DAE residual functor
+ * @tparam T_yy0 Type of initial state
+ * @tparam T_yp0 Type of initial state derivatives
+ * @tparam T_Args Types of pass-through parameters
+ *
+ * @param function_name Calling function name (for printing debugging messages)
```

Comment:
Since user needs to provide __consistent__ initial conditions, i.e. at t=t0, `y'` and `y` need to work together to ensure `f(t0, y, y')=0`, at this point it's user's responsibility.

Reply:
Sorry, I don't quite follow.

Reply:
Unlike ODE, a DAE's initial condition requires both `y(t0)` and `y'(t0)` such that `h(y'(t0), y(t0), t0)=0` for residual `h(y', y, t)`. So for DAE solver user needs to provide consistent `y(t0)` and `y'(t0)` at initial time `t0`. Here "consistency" means in general `y(t0)` and `y'(t0)` cannot be arbitrarily given but must work together to satisfy `h(y'(t0), y(t0), t0)=0`. Therefore Stan's `dae` function requires argument `yy0` for `y(t0)` and argument `yp0` for `y'(t0)`.

Reply:
Yes, I understand this, and the role of `yy0` and `yp0`. I still don't see what the role of the `function_name` argument is.

Reply:
Sorry I misread the comment. We do [use it in ODEs](https://github.com/stan-dev/math/blob/e15c69f777647c0884ab1ac807c9be9101cef641/stan/math/rev/functor/cvodes_integrator.hpp#L211-L233). Same as here, it is part of output error string to differentiate  `dae` vs `dae_tol`, `ode_bdf` vs `ode_bdf_tol`, etc.

Reply:
Got it. Thx.